unittests/Dialect/SPIRV/Runtime/RuntimeTests.cpp

```diff
@@ -0,0 +1,235 @@
+    "spv.globalVariable @var2 bind(1, 0) : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+    "spv.globalVariable @var1 bind(0, 1) : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+    "spv.globalVariable @var0 bind(0, 0) : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+    "spv.globalVariable @globalInvocationID built_in(\"GlobalInvocationId\"): !spv.ptr<vector<3xi32>, Input>\n"
+    "func @kernel() -> () {\n"
+      "%c0 = spv.constant 0 : i32\n"
+
+      "%0 = spv._address_of @var0 : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+      "%1 = spv._address_of @var1 : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+      "%2 = spv._address_of @var2 : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+      "%3 = spv._address_of @var3 : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+
+      "%ptr_id = spv._address_of @globalInvocationID: !spv.ptr<vector<3xi32>, Input>\n"
+
+      "%id = spv.AccessChain %ptr_id[%c0] : !spv.ptr<vector<3xi32>, Input>\n"
+      "%index = spv.Load \"Input\" %id: i32\n"
+
+      "%4 = spv.AccessChain %0[%c0, %index] : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+      "%5 = spv.AccessChain %1[%c0, %index] : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+      "%6 = spv.AccessChain %2[%c0, %index] : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+      "%7 = spv.AccessChain %3[%c0, %index] : !spv.ptr<!spv.struct<!spv.array<1024 x f32 [4]> [0]>, StorageBuffer>\n"
+
+      "%8 = spv.Load \"StorageBuffer\" %4 : f32\n"
+      "%9 = spv.Load \"StorageBuffer\" %5 : f32\n"
+      "%10 = spv.Load \"StorageBuffer\" %6 : f32\n"
+
+      "%11 = spv.FMul %8, %9 : f32\n"
+      "%12 = spv.FAdd %11, %10 : f32\n"
+
+      "spv.Store \"StorageBuffer\" %7, %12 : f32\n"
+      "spv.Return\n"
+    "}\n"
+    "spv.EntryPoint \"GLCompute\" @kernel, @globalInvocationID\n"
+    "spv.ExecutionMode @kernel \"LocalSize\", 1, 1, 1\n"
+"} attributes {\n"
+  "capabilities = [\"Shader\"],\n"
+  "extensions = [\"SPV_KHR_storage_buffer_storage_class\"]\n"
+"}\n";
+
+  createResourceVarFloat(0, 0, 1024);
+  createResourceVarFloat(0, 1, 1024);
+  createResourceVarFloat(1, 0, 1024);
+  createResourceVarFloat(1, 1, 1024);
+
+  auto fmulResult = FMul(vars[0][0], vars[0][1]);
+  auto expected = FAdd(vars[1][0], fmulResult);
+
+  NumWorkGroups numWorkGroups;
+  numWorkGroups.x = 1024;
+  ASSERT_TRUE(succeeded(parseAndRunModule(spirvModuleSource, numWorkGroups)));
+  ASSERT_TRUE(isEqualFloat(expected, vars[1][1]));
+
+  destroyResourceVarFloat(vars[0][0]);
+  destroyResourceVarFloat(vars[0][1]);
+  destroyResourceVarFloat(vars[1][0]);
+  destroyResourceVarFloat(vars[1][1]);
+  destroyResourceVarFloat(fmulResult);
+  destroyResourceVarFloat(expected);
+}
```

Comment:
Can we avoid C++ unit-tests and use a regular lit+FileCheck testing please? I think that is the whole point of using a runner.

See the cuda testing for reference.

Reply:
I think this was just a temporary step to have functionalities implemented incrementally. We don't have all the [shims](https://github.com/tensorflow/mlir/tree/master/tools/mlir-cuda-runner) and [conversions](https://github.com/tensorflow/mlir/blob/master/lib/Conversion/GPUToCUDA/ConvertLaunchFuncToCudaCalls.cpp) like the CUDA side yet. We've already >1k LOCs here; having all the above, we are expecting another >1k LOCs. So I'm fine of landing this as-is. But if you are very uncomfortable about this, I see two ways going forward: 1) remove the C++ tests and land the functionality; 2) leaving this PR open and have following-up commits to implement the rest of the functionality so finally we can squash the intermediate state away.

Reply:
I'm still on the same opinion as in September: build the right tool step by step with the right testing at every step along the way.